openlibrary/plugins/upstream/utils.py

```diff
@@ -709,51 +709,63 @@ def strip_accents(s: str) -> str:
 
 
 @functools.cache
-def get_languages() -> dict:
-    keys = web.ctx.site.things({"type": "/type/language", "limit": 1000})
+def get_languages(limit: int = 1000) -> dict:
+    keys = web.ctx.site.things({"type": "/type/language", "limit": limit})
     return {lang.key: lang for lang in web.ctx.site.get_many(keys)}
 
 
-def autocomplete_languages(prefix: str) -> Iterator[Storage]:
+def _normalize(s: str) -> str:
+    return strip_accents(s).lower()
+
+
+def _matches_lang_name(
+    word_prefix: str, language, translation: str | None = None
+) -> Storage | None:
+    lang_name = language.name
+    if translation is not None:
+        lang_name = safeget(lambda: language['name_translated'][translation][0])
+    # Compare to each inner word of language name for more accurate matching
+    # Eg. for prefix 'greek' will match with 'Ancient Greek' as well as 'Greek'
+    for inner in lang_name.split(' '):
+        if inner and _normalize(inner).startswith(word_prefix):
+            return Storage(
+                key=language.key,
+                code=language.code,
+                name=lang_name,
+            )
+    return None
+
+
+def autocomplete_languages(prefix: str, limit: int) -> Iterator[Storage]:
```

Comment:
Don't need this `limit` here ; can use generators/itertools.islice to avoid keeping track of the number matches.

```suggestion
def autocomplete_languages(prefix: str) -> Iterator[Storage]:
```